src/cargo/sources/git/utils.rs

```diff
@@ -707,6 +742,12 @@ fn reset(repo: &git2::Repository, obj: &git2::Object<'_>, config: &Config) -> Ca
     opts.progress(|_, cur, max| {
         drop(pb.tick(cur, max, ""));
     });
+    if obj.as_tree().is_some() {
+        debug!("doing reset for Cargo nonstandard shallow clone");
+        repo.checkout_tree(obj, Some(&mut opts))?;
+        debug!("reset done");
+        return Ok(());
+    }
     debug!("doing reset");
     repo.reset(obj, git2::ResetType::Hard, Some(&mut opts))?;
```

Comment:
nit: Personally, I feel like an `else {}` would make this code clearer.  While I know early returns are great for reducing code shifting right, I feel like they work best when handling boiler plate.  When there is mutually exclusive business logic, that intent can be made more clear by having separate blocks.
(somehow I originally had missed the `return` and was originally going to ask why both resets happen)